### PR TITLE
fix: strip comments from JSON request bodies in CLI

### LINKED ISSUE #5698: [bug]: Json body serialize in CLI content-type
### Is there an existing issue for this?

- [x] I have searched existing issues and this bug hasn't been reported yet

### Platform

Desktop App

### Browser

Chrome

### Operating System

macOS

### Bug Description

We use AWS Cognito for user authentication.
In desktop we have request to do sign in and in post-request script capture session token to be used in other requests to our API. Works as it should.
Now setting up CLI, export collection and get error. 
AWS use header:
Content-Type: application/x-amz-json-1.1
This is required (and kinda odd...)
Hoppscotch CLI does not recognize the header and wont serialize correctly. Leading to bad data being sent.
Works in desktop so should work the same.

### Deployment Type

Hoppscotch Cloud

### Version

_No response_

### PR DESCRIPTION
**Pull request by jamesgeorge007**
This PR fixes collections with JSON comments failing in CLI with `SerializationException` while working fine in the app, where comments are stripped before sending requests, while the CLI was sending them as-is, breaking APIs like AWS Cognito that expect valid JSON.

Closes #5698, FE-1109.

### What's changed

- Added JSONC support to match app behaviour.
- Integrated `jsonc-parser` library (same as used in `@hoppscotch/common`).
- Created `stripComments()` utility in `src/utils/jsonc.ts` that removes both inline (`//`) and multiline (`/* */`) comments, plus trailing commas.
- Integrated comment stripping in `src/utils/pre-request.ts` before axios JSON serialisation for all JSON content types (`application/json`, `+json`, `-json` variants like `application/x-amz-json-1.1`).
- Added comprehensive unit test suite with 16 test cases covering inline comments, multiline comments, trailing commas, edge cases, and nested structures.
- Added e2e test case in `test.spec.ts` to verify JSONC handling and created a dedicated e2e test fixture collection `jsonc-body-coll.json` with three test scenarios (inline comments, multiline comments, trailing commas).
- Fixed an issue under the `@hoppscotch/common` package as part of the `convertNodeToJSON()` function, where `JSON.stringify(keyNode)` was causing circular reference errors that have been corrected to use `keyNode.value`.

The fix uses `jsonc-parser`'s `parseTree()` with `allowTrailingComma` option to build an AST, then recursively converts it to clean JSON, handling both comments and trailing commas. If parsing fails, it gracefully falls back to the original body string. The stripping occurs in the JSON content-type processing logic in `pre-request.ts`, immediately before Axios serialisation, ensuring custom JSON types (like AWS's `application/x-amz-json-1.1`) work correctly.

### Notes to reviewers

Import the collection [here](https://github.com/user-attachments/files/24584142/cli-validate-col.json) into the app that'd help simulate the concerns mentioned in #5698 and attempt to run it against the CLI. Observe the test failures in the pre-existing app and CLI versions.

The test fixture collection addition can also be used to validate the behaviour. Please feel free to try out other combinations.

### CODE REVIEW

File: packages/hoppscotch-cli/src/utils/jsonc.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+ * @returns The clean JSON string without comments or trailing commas.
+ */
+export function stripComments(jsoncString: string): string {
+  // Handle empty or whitespace-only strings
+  if (!jsoncString || jsoncString.trim() === "") {
+    return jsoncString;
+  }
+
+  // First strip basic comments, then handle trailing commas
+  return stripCommentsAndCommas(stripComments_(jsoncString) ?? jsoncString);
```
[the remark this answers is not in the document]
Comment (jamesgeorge007) [reply packages/hoppscotch-cli/src/utils/jsonc.ts:88]: Added null-coalescing safeguard (`??` `jsonString`) to `common` package. Both CLI and the `common` package now have consistent defensive programming against potential null returns from jsonc-parser.

File: packages/hoppscotch-cli/src/__tests__/unit/jsonc.spec.ts [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  describe("handles null return from stripComments_", () => {
+    test("gracefully handles potential null from jsonc-parser", () => {
+      const input = '{"key": "value"}';
+      const result = stripComments(input);
+      expect(result).toBeTruthy();
+      const parsed = JSON.parse(result);
+      expect(parsed).toEqual({ key: "value" });
+    });
+  });
+});
```
[the remark this answers is not in the document]
Comment (jamesgeorge007) [reply packages/hoppscotch-cli/src/__tests__/unit/jsonc.spec.ts:151]: Added 3 unit test cases for malformed JSON behavior. Tests verify that jsonc-parser attempts to repair malformed JSON when possible, and the function handles completely invalid input gracefully. Note: jsonc-parser is lenient and tries to fix JSON rather than returning original text unchanged.

File: packages/hoppscotch-cli/src/utils/pre-request.ts [hunk 3]
```diff
@@ -578,6 +581,55 @@ function getFinalBodyFromRequest(
...
+      // If parsing fails after stripping comments, return error to provide
+      // immediate feedback instead of sending invalid JSON to the API.
+      // We include the cleaned body so the error context reflects what was
+      // actually parsed (after env substitution and comment stripping).
+      return E.left(
+        error({
+          code: "PARSING_ERROR",
+          data: `${cleanedBody} (Invalid JSON in request body: ${err instanceof Error ? err.message : String(err)})`,
+        })
+      );
```
[the remark this answers is not in the document]
Comment (jamesgeorge007) [reply packages/hoppscotch-cli/src/utils/pre-request.ts]: Replaced the word boundary regex with precise MIME type suffix matching. Now extracts and normalises the MIME type (removing parameters) and checks for exact match of application/json or endings with `+json`, `/json`, or `-json`. This prevents false positives like text/jsonify while correctly handling all valid JSON content types, including the AWS Cognito `application/x-amz-json-1.1` case.

File: packages/hoppscotch-cli/src/__tests__/e2e/fixtures/collections/jsonc-body-coll.json [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    }
+  ],
+  "auth": {
+    "authType": "inherit",
+    "authActive": true
+  },
+  "headers": [],
+  "variables": [],
+  "description": ""
+}
```
[the remark this answers is not in the document]
Comment (jamesgeorge007) [reply packages/hoppscotch-cli/src/__tests__/e2e/fixtures/collections/jsonc-body-coll.json:74]: Unnecessary, given this is just another use case where the content type overriding would happen via header,s given the preset list doesn't recognise the `application/x-amz-json-1.1` header.

https://github.com/hoppscotch/hoppscotch/blob/b3b609d0b420266411abf5dd4c5b76a1c6334eae/packages/hoppscotch-data/src/rest/v/9/body.ts#L37-L66

File: packages/hoppscotch-cli/src/__tests__/unit/jsonc.spec.ts [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+      const result = stripComments(input);
+      // The parser will attempt to close the brace
+      expect(result).toBe('{"key":"value"}');
+    });
+
+    test("gracefully handles completely invalid JSON", () => {
+      const input = 'this is not json at all {]}{]';
+      const result = stripComments(input);
+      // jsonc-parser extracts what it can and returns an object (even if mostly empty)
+      expect(result).toBe('{}');
```
[the remark this answers is not in the document]
Comment (jamesgeorge007) [reply packages/hoppscotch-cli/src/__tests__/unit/jsonc.spec.ts:141]: These assertions are actually correct. Through empirical testing (all 19 unit tests pass), `jsonc-parser` demonstrates lenient behaviour and attempts to repair malformed JSON. The parser successfully converts `{"key": "value"` to `{"key": "value"}` and `this is not json at all {]}{]'` to `'{}'`. The tests accurately reflect the library's actual behaviour.

File: packages/hoppscotch-cli/src/utils/pre-request.ts [hunk 6]
```diff
@@ -578,6 +581,59 @@ function getFinalBodyFromRequest(
...
+      return E.right(JSON.stringify(parsedBody));
+    } catch (err) {
+      // If parsing fails after stripping comments, return error to provide
+      // immediate feedback instead of sending invalid JSON to the API.
+      // We include the cleaned body so the error context reflects what was
+      // actually parsed (after env substitution and comment stripping).
+      return E.left(
+        error({
+          code: "PARSING_ERROR",
+          data: `${cleanedBody} (Invalid JSON in request body: ${err instanceof Error ? err.message : String(err)})`,
```
[the remark this answers is not in the document]
Comment (jamesgeorge007) [reply packages/hoppscotch-cli/src/utils/pre-request.ts]: Changed error message to use `request.body.body` (original template string) instead of `cleanedBody` (with resolved environment variables). This prevents leaking sensitive values like API keys, passwords, and tokens that were substituted from environment variables. Thank you for catching this!

File: packages/hoppscotch-cli/src/__tests__/unit/jsonc.spec.ts [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    });
+  });
+
+  describe("handles malformed JSON", () => {
+    test("attempts to parse malformed JSON and returns result", () => {
+      // jsonc-parser is lenient and tries to repair malformed JSON
+      const input = '{"key": "value"'; // missing closing brace
+      const result = stripComments(input);
+      // The parser will attempt to close the brace
+      expect(result).toBe('{"key":"value"}');
```
[the remark this answers is not in the document]
Comment (jamesgeorge007) [reply packages/hoppscotch-cli/src/__tests__/unit/jsonc.spec.ts:134]: https://github.com/hoppscotch/hoppscotch/pull/5769#discussion_r2687216032

**Review by anwarulislam — APPROVED** [review]
LGTM 🚀
